### PR TITLE
Add tests for EH Namespace in e2e provisioning test suite

### PR DESCRIPTION
**Pull request by sayanh**
**Description**

Changes proposed in this pull request:

- Add EH Namespace test for e2e provisioning

**Related issue(s)**

See https://github.com/kyma-project/kyma/issues/7922

### CODE REVIEW

File: tests/e2e/provisioning/test/e2e_provisioning_test.go [hunk 1]
```diff
@@ -46,4 +49,33 @@ func Test_E2E_Provisioning(t *testing.T) {
...
+	if ts.IsTestAzureEventHubsEnabled {
+		checkAzureEventHubProperties(ts, t, operationID)
+	}
+}
+
+func checkAzureEventHubProperties(ts *Suite, t *testing.T, operationID string) {
+	resourceGroup, err := (*ts.azureClient).GetResourceGroup(context.TODO(), ts.brokerClient.GetClusterName())
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusOK, resourceGroup.StatusCode, "HTTP GET fails for ResourceGroup")
+	assert.NotNil(t, resourceGroup, "ResourceGroup name is nil")
```
Comment (k15r) [inline tests/e2e/provisioning/test/e2e_provisioning_test.go:62]: its a struct, not a pointer to a struct, so it can never be nil, and your not checking the name here

File: tests/e2e/provisioning/test/e2e_provisioning_test.go [hunk 2]
```diff
@@ -46,4 +49,33 @@ func Test_E2E_Provisioning(t *testing.T) {
...
+	assert.Equal(t, ts.brokerClient.GetClusterName(), *resourceGroup.Name, "ResourceGroup Name is incorrect")
+
+	// Check tags for ResourceGroup
+	assert.Equal(t, operationID, *resourceGroup.Tags[azure.TagOperationID], "Tag OperationID for ResourceGroup is incorrect")
+	assert.Equal(t, ts.brokerClient.InstanceID(), *resourceGroup.Tags[azure.TagInstanceID], "Tag InstanceID for ResourceGroup is incorrect")
+	assert.Equal(t, ts.brokerClient.SubAccountID(), *resourceGroup.Tags[azure.TagSubAccountID], "Tag SubAccountID for ResourceGroup is incorrect")
+
+	namespace, err := (*ts.azureClient).GetEHNamespace(context.TODO(), ts.brokerClient.GetClusterName(), ts.brokerClient.GetClusterName())
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusOK, resourceGroup.StatusCode, "HTTP GET fails for EH Namespace")
```
Comment (k15r) [inline tests/e2e/provisioning/test/e2e_provisioning_test.go:72]: StatusOk == 200
But all 2xx Codes would indicate a success. So somehow the message is off

File: tests/e2e/provisioning/test/e2e_provisioning_test.go [hunk 3]
```diff
@@ -46,4 +49,33 @@ func Test_E2E_Provisioning(t *testing.T) {
...
+	assert.NotNil(t, namespace, "EH Namespace name is nil")
```
Comment (k15r) [inline tests/e2e/provisioning/test/e2e_provisioning_test.go:73]: Your not checking the namespaces name here

Comment (k15r) [reply tests/e2e/provisioning/test/e2e_provisioning_test.go:73]: require would be the better option, as later asserts will try to dereference nil otherwise

File: tests/e2e/provisioning/internal/hyperscaler/account_pool.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+		secret.Labels["tenantName"] = tenantName
+		updatedSecret, err := p.secretsClient.Update(secret)
+		if err != nil {
+			return Credentials{}, errors.Wrapf(err, "accountPool error while updating secret with tenantName: %s", tenantName)
+		}
+		return credentialsFromSecret(updatedSecret, hyperscalerType, tenantName), nil
+	}
+
+	return Credentials{}, fmt.Errorf("accountPool failed to find unassigned secret for hyperscalerType: %s",
+		hyperscalerType)
```
[the remark this answers is not in the document]
Comment (sayanh) [reply tests/e2e/provisioning/internal/hyperscaler/account_pool.go]: done

File: tests/e2e/provisioning/internal/hyperscaler/account_pool.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	})
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "accountPool error during secret list for LabelSelector: %s", labelSelector)
+	}
+
+	if secrets != nil && len(secrets.Items) < 1 {
+		return nil, errors.Wrapf(err, "no secrets returned for LabelSelector: %s", labelSelector)
+	}
+	return &secrets.Items[0], nil
```
[the remark this answers is not in the document]
Comment (sayanh) [reply tests/e2e/provisioning/internal/hyperscaler/account_pool.go]: makes sense, checking first and then accessing

File: tests/e2e/provisioning/internal/hyperscaler/azure/client.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+type AzureInterface interface {
+	GetResourceGroup(ctx context.Context, name string) (resources.Group, error)
+	GetEHNamespace(ctx context.Context, resourceGroupName, namespaceName string) (eventhub.EHNamespace, error)
+}
+
+var _ AzureInterface = (*AzureClient)(nil)
+
+type AzureClient struct {
+	eventhubNamespaceClient eventhub.NamespacesClient
+	resourcegroupClient     resources.GroupsClient
```
[the remark this answers is not in the document]
Comment (sayanh) [reply tests/e2e/provisioning/internal/hyperscaler/azure/client.go]: changed to `resourceGroupClient` and `eventHubNamespaceClient`

File: tests/e2e/provisioning/internal/hyperscaler/azure/env.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	autorestazure "github.com/Azure/go-autorest/autorest/azure"
+
+	"github.com/kyma-incubator/compass/tests/e2e/provisioning/internal/hyperscaler"
+)
+
+// GetConfig returns Azure config
+func GetConfig(clientID, clientSecret, tenantID, subscriptionID, location string) (*Config, error) {
+	config := NewDefaultConfig()
+	azureEnv, err := autorestazure.EnvironmentFromName("AzurePublicCloud") // shouldn't fail
```
[the remark this answers is not in the document]
Comment (sayanh) [reply tests/e2e/provisioning/internal/hyperscaler/azure/env.go]: removed

File: tests/e2e/provisioning/internal/hyperscaler/azure/env.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+	config.clientID = clientID
+	config.clientSecret = clientSecret
+	config.tenantID = tenantID
+	config.subscriptionID = subscriptionID
+	config.location = location
+	return config, nil
+}
+
+// GetConfigFromHAPCredentialsAndProvisioningParams returns Azure config from HAPCredentials
+func GetConfigFromHAPCredentialsAndProvisioningParams(credentials hyperscaler.Credentials, region string) (*Config, error) {
```
[the remark this answers is not in the document]
Comment (sayanh) [reply tests/e2e/provisioning/internal/hyperscaler/azure/env.go:51]: right, added checks now

File: tests/e2e/provisioning/internal/hyperscaler/azure/provider.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+		return nil, errors.Wrap(err, "while creating resource group client")
+	}
+
+	// create azure client
+	return NewAzureClient(nsClient, resourcegroupClient), nil
+}
+
+// getNamespaceClient returns a client for handling of Azure Namespaces
+func (ac *azureProvider) getNamespaceClient(config *Config, authorizer autorest.Authorizer) (eventhub.NamespacesClient, error) {
+	nsClient := eventhub.NewNamespacesClient(config.subscriptionID)
```
[the remark this answers is not in the document]
Comment (sayanh) [reply tests/e2e/provisioning/internal/hyperscaler/azure/provider.go:59]: added checks

File: tests/e2e/provisioning/internal/hyperscaler/azure/provider.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	oauthConfig, err := adal.NewOAuthConfig(environment.ActiveDirectoryEndpoint, config.tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := adal.NewServicePrincipalToken(*oauthConfig, config.clientID, config.clientSecret, environment.ResourceManagerEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	return autorest.NewBearerAuthorizer(token), err
```
[the remark this answers is not in the document]
Comment (sayanh) [reply tests/e2e/provisioning/internal/hyperscaler/azure/provider.go]: changed to `nil`

File: tests/e2e/provisioning/test/e2e_provisioning_test.go (hunk 1 above)
Comment (sayanh) [reply tests/e2e/provisioning/test/e2e_provisioning_test.go]: This is changed to listing of resourcegroups hence this is not applicable anymore.

Name will be a generated one by KEB hence not checking it at all(in the new code).

File: tests/e2e/provisioning/test/e2e_provisioning_test.go (hunk 2 above)
Comment (sayanh) [reply tests/e2e/provisioning/test/e2e_provisioning_test.go]: I expect to get a 200 only hence 2xx does not make sense here.

File: tests/e2e/provisioning/test/e2e_provisioning_test.go (hunk 3 above)
Comment (sayanh) [reply tests/e2e/provisioning/test/e2e_provisioning_test.go]: Again, checking the name of the NS is not critical here.

File: tests/e2e/provisioning/test/e2e_provisioning_test.go [hunk 11]
```diff
@@ -46,4 +50,39 @@ func Test_E2E_Provisioning(t *testing.T) {
...
+
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusOK, groupListResultPage.Response().StatusCode, "HTTP GET fails for ListResourceGroup")
+	assert.Equal(t, 1, len(groupListResultPage.Values()), "groupListResultPage should return 1 ResourceGroup")
+
+	rg := groupListResultPage.Values()[0]
+	tagsRG := groupListResultPage.Values()[0].Tags
+
+	// Check tags for ResourceGroup
+	assert.Equal(t, operationID, *tagsRG[azure.TagOperationID], "Tag OperationID for ResourceGroup is incorrect")
```
Comment (nachtmaar) [inline tests/e2e/provisioning/test/e2e_provisioning_test.go:71]: Edge case: what if the key is not inside the map. Then nil will be dereferenced => can we add a check for nil ? :) Same for other map tag keys

**Comment by sayanh** [comment]
Code duplication for internal folders will be addressed [here](https://github.com/kyma-incubator/compass/pull/1295)

File: tests/e2e/provisioning/test/e2e_provisioning_test.go [hunk 12]
```diff
@@ -46,4 +50,39 @@ func Test_E2E_Provisioning(t *testing.T) {
...
+	assert.Equal(t, ts.brokerClient.InstanceID(), *tagsRG[azure.TagInstanceID], "Tag InstanceID for ResourceGroup is incorrect")
+	assert.Equal(t, ts.brokerClient.SubAccountID(), *tagsRG[azure.TagSubAccountID], "Tag SubAccountID for ResourceGroup is incorrect")
+
+	ehNamespaceListResultPage, err := (*ts.azureClient).ListEHNamespaceByResourceGroup(context.TODO(), *rg.Name)
```
Comment (nachtmaar) [inline tests/e2e/provisioning/test/e2e_provisioning_test.go:75]: Probably `rg.Name` will never be nill if a successful response returned, but can we still check for nil here? :)

File: tests/e2e/provisioning/internal/hyperscaler/account_pool.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+	"sync"
+
+	"github.com/pkg/errors"
+
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type HyperscalerType string
```
Comment (marcobebway) [inline tests/e2e/provisioning/internal/hyperscaler/account_pool.go:14]: (nit) Hyperscaler is redundant, it reads `hyperscaler.HyperscalerType`.

File: tests/e2e/provisioning/internal/hyperscaler/azure/client.go [hunk 14]
```diff
@@ -0,0 +1,35 @@
+package azure
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/eventhub/mgmt/2017-04-01/eventhub"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
+)
+
+type AzureInterface interface {
```
Comment (marcobebway) [inline tests/e2e/provisioning/internal/hyperscaler/azure/client.go:10]: (nit) Azure is redundant, it reads `azure.AzureInterface`.

File: tests/e2e/provisioning/README.md [hunk 15]
```diff
@@ -72,5 +72,6 @@ You can configure the test execution by using the following environment variable
 | **APP_SKIP_CERT_VERIFICATION** | Specifies whether TLS checks the presented certificates. | `false` |
 | **APP_DUMMY_TEST** | Specifies if test should success without any action. | `false` |
 | **APP_CLEANUP_PHASE** | Specifies if the test executes the cleanup phase. | `false` |
+| **APP_TEST_AZURE_EVENTHUBS_ENABLED** | Specifies if the test should test the infrastructure in Azure Event Hubs. | `true` |
```
Comment (marcobebway) [inline tests/e2e/provisioning/README.md:75]: Where `APP_TEST_AZURE_EVENTHUBS_ENABLED` is used?

File: tests/e2e/provisioning/internal/hyperscaler/account_pool.go [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+const (
+	Azure HyperscalerType = "azure"
+)
+
+type Credentials struct {
+	CredentialName  string
```
Comment (marcobebway) [inline tests/e2e/provisioning/internal/hyperscaler/account_pool.go:21]: (nit) probably `Name` is enough?

**Comment by nachtmaar** [comment]
> Code duplication for internal folders will be addressed [here](https://github.com/kyma-incubator/compass/pull/1295)

@sayanh You referenced the wrong issue. This is the correct one: https://github.com/kyma-incubator/compass/issues/1303

File: tests/e2e/provisioning/pkg/client/broker/broker_client.go [hunk 17]
```diff
@@ -132,6 +135,22 @@ func (c *Client) DeprovisionRuntime() (string, error) {
...
+
+func (c *Client) InstanceID() string {
+	return c.instanceID
+}
+
+func (c *Client) SubAccountID() string {
+	return c.subAccountID
+}
+
+func (c *Client) GetClusterName() string {
```
Comment (marcobebway) [inline tests/e2e/provisioning/pkg/client/broker/broker_client.go:150]: Why `Get` in the function name? Notice other getter functions does not follow this pattern?

File: tests/e2e/provisioning/test/e2e_provisioning_test.go [hunk 18]
```diff
@@ -46,4 +50,39 @@ func Test_E2E_Provisioning(t *testing.T) {
 
 	err = ts.dashboardChecker.AssertRedirectedToUAA(dashboardURL)
 	assert.NoError(t, err)
+
+	if ts.IsTestAzureEventHubsEnabled {
+		checkAzureEventHubProperties(ts, t, operationID)
+	}
+}
+
+func checkAzureEventHubProperties(ts *Suite, t *testing.T, operationID string) {
```
Comment (marcobebway) [inline tests/e2e/provisioning/test/e2e_provisioning_test.go:59]: Does it make sense to check the provisioning state for the resource group and namespace?

File: tests/e2e/provisioning/test/e2e_provisioning_test.go [hunk 19]
```diff
@@ -46,4 +50,39 @@ func Test_E2E_Provisioning(t *testing.T) {
 
 	err = ts.dashboardChecker.AssertRedirectedToUAA(dashboardURL)
 	assert.NoError(t, err)
+
+	if ts.IsTestAzureEventHubsEnabled {
+		checkAzureEventHubProperties(ts, t, operationID)
```
Comment (polskikiel) [inline tests/e2e/provisioning/test/e2e_provisioning_test.go:55]: Please add here info log that test checks Azure EH properties

File: tests/e2e/provisioning/test/test_suite.go [hunk 20]
```diff
@@ -147,6 +171,20 @@ func (ts *Suite) Cleanup() {
 	require.NoError(ts.t, err)
 	err = ts.brokerClient.AwaitOperationSucceeded(operationID, ts.DeprovisionTimeout)
 	assert.NoError(ts.t, err)
+
+	if ts.IsTestAzureEventHubsEnabled {
+		ts.ensureAzureResourceGroupRemoved()
```
Comment (polskikiel) [inline tests/e2e/provisioning/test/test_suite.go:176]: you can add a similar log here

File: tests/e2e/provisioning/test/e2e_provisioning_test.go (hunk 18 above)
Comment (sayanh) [reply tests/e2e/provisioning/test/e2e_provisioning_test.go:59]: No, cause this step is done after Kyma runtime is provisioned. So by this time, Azure event hubs should definitely be created. Waiting should not be necessary.

Comment (marcobebway) [reply tests/e2e/provisioning/test/e2e_provisioning_test.go:59]: Correct, waiting does not make sense at that point. 

What about checking the provisioning state for the resource group and namespace to be successful? To elaborate more, Is the provisioning state for resource group and namespace reflected somehow as part of the Kyma runtime provisioning state? This is not clear to me.

File: tests/e2e/provisioning/test/e2e_provisioning_test.go (hunk 11 above)
Comment (sayanh) [reply tests/e2e/provisioning/test/e2e_provisioning_test.go]: checking for "nil"s now ;P

File: tests/e2e/provisioning/test/e2e_provisioning_test.go (hunk 12 above)
Comment (sayanh) [reply tests/e2e/provisioning/test/e2e_provisioning_test.go:82]: checking for nil now

File: tests/e2e/provisioning/internal/hyperscaler/account_pool.go (hunk 13 above)
Comment (sayanh) [reply tests/e2e/provisioning/internal/hyperscaler/account_pool.go]: Changed to `Type`

File: tests/e2e/provisioning/internal/hyperscaler/azure/client.go (hunk 14 above)
Comment (sayanh) [reply tests/e2e/provisioning/internal/hyperscaler/azure/client.go]: changed to `Interface`

File: tests/e2e/provisioning/README.md (hunk 15 above)
Comment (sayanh) [reply tests/e2e/provisioning/README.md]: changed

File: tests/e2e/provisioning/internal/hyperscaler/account_pool.go (hunk 16 above)
Comment (sayanh) [reply tests/e2e/provisioning/internal/hyperscaler/account_pool.go]: changed to `Name`

File: tests/e2e/provisioning/pkg/client/broker/broker_client.go (hunk 17 above)
Comment (sayanh) [reply tests/e2e/provisioning/pkg/client/broker/broker_client.go]: removed `Get`

File: tests/e2e/provisioning/test/e2e_provisioning_test.go (hunk 19 above)
Comment (sayanh) [reply tests/e2e/provisioning/test/e2e_provisioning_test.go:56]: Added logs for this step

File: tests/e2e/provisioning/test/test_suite.go (hunk 20 above)
Comment (sayanh) [reply tests/e2e/provisioning/test/test_suite.go:177]: Added logs for this step
